File: Leetcode/Java Codes/October/_15.java
import java.util.*;

class _15 {
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];

        for(int i=0; i<n; i++){
            arr[i] = sc.nextInt();
        }

        List<Integer> nums = new ArrayList<>();
        for(int a : arr){
            nums.add(a);
        }

        System.out.print(helper(nums));
        sc.close();
    }    

    static int helper(List<Integer> nums){
        int n = nums.size();
        //int k = n/2;
        int ans = 0;

        int[] inc = new int[n];
        inc[0] = 1;
        for(int i=1; i<n; i++){
            if(nums.get(i)>nums.get(i-1)){
                inc[i] = inc[i-1]+1;
            } else {
                inc[i] = 1;
            }
        }

        int left=0, right = n/2;
        while(left<=right){
            int mid = left + (right-left)/2;
            if(isTrue(nums,inc,mid)){
                ans=mid;
                left=mid+1;
            }
            else right = mid-1;
        }

        return ans;
    }

    static boolean isTrue(List<Integer> nums, int[] inc, int k) {
        int n = nums.size();
        for (int i = k - 1; i + k < n; i++) {
            if (inc[i] >= k && inc[i + k] >= k) return true;
        }
        return false;
    }
}

File: Leetcode/Python/October/_15.py
class Solution:
    def isTrue(self, arr, k, inc) -> bool:
        n = len(arr)
        for i in range(k-1, n-k):
            if inc[i] >= k and inc[i+k] >= k:
                return True
        return False

    def helper(self, arr) -> int:
        n = len(arr)
        inc = [1] * n  
    
        for i in range(1, n):
            if arr[i] > arr[i-1]:
                inc[i] = inc[i-1] + 1
            else:
                inc[i] = 1  
        
        left, right = 1, n//2
        ans = 0
        
        while left <= right:
            mid = left + (right-left)//2
            if self.isTrue(arr, mid, inc):
                ans = mid
                left = mid + 1
            else:
                right = mid - 1
                
        return ans  

def main():
    arr = list(map(int, input().split()))
    sol = Solution()
    print(sol.helper(arr))

if __name__ == "__main__":
    main()